src/tasks/PageIterator.ts

```diff
@@ -0,0 +1,332 @@
+   * Member holding the current position on the collection
+   */
+  private cursor: number;
+
+  /**
+   * @private
+   * Member holding the factory to create the parsable object
+   */
+  private readonly parsableFactory: ParsableFactory<PageCollection<T>>;
+
+  /**
+   * @private
+   * Member holding the error mappings
+   */
+  private readonly errorMappings: ErrorMappings;
+
+  /**
+   * @private
+   * Member holding the callback for iteration
+   */
+  private readonly callback: PageIteratorCallback<T>;
+
+  /**
+   * @private
+   * Member holding the state of the iterator
+   */
+  private pagingState: PagingState;
+
+  /**
+   * @private
+   * Member holding the headers for the request
+   */
+  private readonly options?: PagingRequestOptions;
+
+  /**
+   * @public
+   * @constructor
+   * Creates new instance for PageIterator
+   * @returns An instance of a PageIterator
+   * @param requestAdapter - The request adapter
+   * @param pageResult - The page collection result of T
+   * @param callback - The callback function to be called on each item
+   * @param errorMappings - The error mappings
+   * @param parsableFactory - The factory to create the parsable object collection
+   * @param options - The request options to configure the request
+   */
+  public constructor(
+    requestAdapter: RequestAdapter,
+    pageResult: PageCollection<T>,
+    callback: PageIteratorCallback<T>,
+    parsableFactory: ParsableFactory<PageCollection<T>>,
+    errorMappings: ErrorMappings,
+    options?: PagingRequestOptions,
+  ) {
+    if (!requestAdapter) {
+      const error = new Error("Request adapter is undefined, Please provide a valid request adapter");
+      error.name = "Invalid Request Adapter Error";
+      throw error;
+    }
+    if (!pageResult) {
+      const error = new Error("Page result is undefined, Please provide a valid page result");
+      error.name = "Invalid Page Result Error";
+      throw error;
+    }
+    if (!callback) {
+      const error = new Error("Callback is undefined, Please provide a valid callback");
+      error.name = "Invalid Callback Error";
+      throw error;
+    }
+    if (!parsableFactory) {
+      const error = new Error("Parsable factory is undefined, Please provide a valid parsable factory");
+      error.name = "Invalid Parsable Factory Error";
+      throw error;
+    }
+    if (!errorMappings) {
+      const error = new Error("Error mappings is undefined, Please provide a valid error mappings");
+      error.name = "Invalid Error Mappings Error";
+      throw error;
+    }
+    this.requestAdapter = requestAdapter;
+    this.currentPage = pageResult;
+
+    this.cursor = 0;
+    this.complete = false;
+    this.errorMappings = errorMappings;
+    this.parsableFactory = parsableFactory;
+    this.callback = callback;
+
+    if (!options) {
+      options = {};
+    }
+    if (!options.headers) {
+      options.headers = new Headers();
+    }
+    if (!options.headers.has("Content-Type")) {
```

Comment:
I think since you'll be typically doing a GET request to get the next page, setting the `Content-Type` header could arguably be incorrect/unnecessary